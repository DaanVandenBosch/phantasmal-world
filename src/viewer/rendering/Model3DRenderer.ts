import {
    AdditiveBlending,
    AnimationClip,
    AnimationMixer,
    Clock,
    DoubleSide,
    Mesh,
    MeshBasicMaterial,
    MeshLambertMaterial,
    Object3D,
    PerspectiveCamera,
    SkeletonHelper,
    SkinnedMesh,
    Vector3,
} from "three";
import { Disposable } from "../../core/observable/Disposable";
import { NjMotion } from "../../core/data_formats/parsing/ninja/motion";
import { xvr_texture_to_texture } from "../../core/rendering/conversion/ninja_textures";
import { create_mesh, create_skinned_mesh } from "../../core/rendering/conversion/create_mesh";
import { ninja_object_to_buffer_geometry } from "../../core/rendering/conversion/ninja_geometry";
import {
    create_animation_clip,
    PSO_FRAME_RATE,
} from "../../core/rendering/conversion/ninja_animation";
import { DisposableThreeRenderer, Renderer } from "../../core/rendering/Renderer";
import { Disposer } from "../../core/observable/Disposer";
import { ChangeEvent } from "../../core/observable/Observable";
import { Model3DStore } from "../stores/Model3DStore";
import { LogManager } from "../../core/Logger";

const logger = LogManager.get("viewer/rendering/Model3DRenderer");

const DEFAULT_MATERIAL = new MeshLambertMaterial({
    color: 0xffffff,
    side: DoubleSide,
});
const DEFAULT_SKINNED_MATERIAL = new MeshLambertMaterial({
    skinning: true,
    color: 0xffffff,
    side: DoubleSide,
});

export class Model3DRenderer extends Renderer implements Disposable {
    private readonly disposer = new Disposer();
    private readonly clock = new Clock();
    private mesh?: Object3D;
    private skeleton_helper?: SkeletonHelper;
    private animation?: {
        mixer: AnimationMixer;
        clip: AnimationClip;
    };
    private update_animation_time = true;

    readonly camera = new PerspectiveCamera(75, 1, 1, 200);

    constructor(
        three_renderer: DisposableThreeRenderer,
        private readonly model_3d_store: Model3DStore,
    ) {
        super(three_renderer);

        this.disposer.add_all(
            model_3d_store.current_nj_data.observe(this.nj_data_or_xvm_changed),
            model_3d_store.current_textures.observe(this.nj_data_or_xvm_changed),
            model_3d_store.current_nj_motion.observe(this.nj_motion_changed),
            model_3d_store.show_skeleton.observe(this.show_skeleton_changed),
            model_3d_store.animation_playing.observe(this.animation_playing_changed),
            model_3d_store.animation_frame_rate.observe(this.animation_frame_rate_changed),
            model_3d_store.animation_frame.observe(this.animation_frame_changed),
        );

        this.init_camera_controls();
    }

    set_size(width: number, height: number): void {
        this.camera.aspect = width / height;
        this.camera.updateProjectionMatrix();
        super.set_size(width, height);
    }

    dispose(): void {
        super.dispose();
        this.disposer.dispose();
    }

    protected render(): void {
        if (this.animation) {
            this.animation.mixer.update(this.clock.getDelta());
        }

        this.light_holder.quaternion.copy(this.camera.quaternion);
        super.render();

        if (this.animation && !this.animation.mixer.clipAction(this.animation.clip).paused) {
            this.update_animation_frame();
            this.schedule_render();
        }
    }

    private nj_data_or_xvm_changed = (): void => {
        if (this.mesh) {
            this.scene.remove(this.mesh);
            this.mesh = undefined;
            this.scene.remove(this.skeleton_helper!);
            this.skeleton_helper = undefined;
        }

        if (this.animation) {
            this.animation.mixer.stopAllAction();
            if (this.mesh) this.animation.mixer.uncacheRoot(this.mesh);
            this.animation = undefined;
        }

        const nj_data = this.model_3d_store.current_nj_data.val;

        if (nj_data) {
            const { nj_object, has_skeleton } = nj_data;

            let mesh: Mesh;

            const textures = this.model_3d_store.current_textures.val.map(tex => {
                if (tex) {
                    try {
                        return xvr_texture_to_texture(tex);
                    } catch (e) {
                        logger.error("Couldn't convert XVR texture.", e);
                    }
                }

                return undefined;
            });

            const materials = textures.map(tex =>
                tex
                    ? new MeshBasicMaterial({
                          skinning: has_skeleton,
                          map: tex,
                          side: DoubleSide,
                          alphaTest: 0.1,
                          transparent: true,
                      })
                    : new MeshLambertMaterial({
                          skinning: has_skeleton,
                          side: DoubleSide,
                      }),
            );

            if (has_skeleton) {
                mesh = create_skinned_mesh(
                    ninja_object_to_buffer_geometry(nj_object),
                    materials,
                    DEFAULT_SKINNED_MATERIAL,
                );
            } else {
                mesh = create_mesh(
                    ninja_object_to_buffer_geometry(nj_object),
                    materials,
                    DEFAULT_MATERIAL,
                );
            }

            // Make sure we rotate around the center of the model instead of its origin.
            const bb = mesh.geometry.boundingBox;
            const height = bb.max.y - bb.min.y;
            mesh.translateY(-height / 2 - bb.min.y);

            this.mesh = mesh;
            this.scene.add(mesh);

            this.skeleton_helper = new SkeletonHelper(mesh);
            this.skeleton_helper.visible = this.model_3d_store.show_skeleton.val;
            (this.skeleton_helper.material as any).linewidth = 3;
            this.scene.add(this.skeleton_helper);

            this.reset_camera(new Vector3(0, 10, 20), new Vector3(0, 0, 0));
        }

        this.schedule_render();
    };

    private nj_motion_changed = ({ value: nj_motion }: ChangeEvent<NjMotion | undefined>): void => {
        let mixer!: AnimationMixer;

        if (this.animation) {
            this.animation.mixer.stopAllAction();
            mixer = this.animation.mixer;
        }

        const nj_data = this.model_3d_store.current_nj_data.val;

        if (!this.mesh || !(this.mesh instanceof SkinnedMesh) || !nj_motion || !nj_data) return;

        if (!this.animation) {
            mixer = new AnimationMixer(this.mesh);
        }

        const clip = create_animation_clip(nj_data.nj_object, nj_motion);

        this.animation = {
            mixer,
            clip,
        };

        this.clock.start();
        this.animation.mixer.clipAction(this.animation.clip).play();
        this.schedule_render();
    };

    private show_skeleton_changed = ({ value: show_skeleton }: ChangeEvent<boolean>): void => {
        if (this.skeleton_helper) {
            this.skeleton_helper.visible = show_skeleton;
            this.schedule_render();
        }
    };

    private animation_playing_changed = ({ value: playing }: ChangeEvent<boolean>): void => {
        if (this.animation) {
            this.animation.mixer.clipAction(this.animation.clip).paused = !playing;

            if (playing) {
                this.clock.start();
                this.schedule_render();
            } else {
                this.clock.stop();
            }
        }
    };

    private animation_frame_rate_changed = ({ value: frame_rate }: ChangeEvent<number>): void => {
        if (this.animation) {
            this.animation.mixer.timeScale = frame_rate / PSO_FRAME_RATE;
        }
    };

    private animation_frame_changed = ({ value: frame }: ChangeEvent<number>): void => {
        const nj_motion = this.model_3d_store.current_nj_motion.val;

        if (this.animation && nj_motion) {
            const frame_count = nj_motion.frame_count;
            if (frame > frame_count) frame = 1;
            if (frame < 1) frame = frame_count;

            if (this.update_animation_time) {
                this.animation.mixer.clipAction(this.animation.clip).time =
                    (frame - 1) / PSO_FRAME_RATE;
            }

            this.schedule_render();
        }
    };

    private update_animation_frame(): void {
        if (this.animation) {
            const action = this.animation.mixer.clipAction(this.animation.clip);

            if (!action.paused) {
                this.update_animation_time = false;
                this.model_3d_store.set_animation_frame(action.time * PSO_FRAME_RATE + 1);
                this.update_animation_time = true;
            }
        }
    }
}
