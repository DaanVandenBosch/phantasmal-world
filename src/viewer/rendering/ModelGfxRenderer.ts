import { ModelStore } from "../stores/ModelStore";
import { Disposer } from "../../core/observable/Disposer";
import { Renderer } from "../../core/rendering/Renderer";
import { GfxRenderer } from "../../core/rendering/GfxRenderer";
import { ninja_object_to_node } from "../../core/rendering/conversion/ninja_geometry";

export class ModelGfxRenderer implements Renderer {
    private readonly disposer = new Disposer();

    readonly canvas_element: HTMLCanvasElement;

    constructor(private readonly store: ModelStore, private readonly renderer: GfxRenderer) {
        this.canvas_element = renderer.canvas_element;

        renderer.camera.pan(0, 0, -50);

        this.disposer.add_all(store.current_nj_object.observe(this.nj_object_or_xvm_changed));
    }

    dispose(): void {
        this.disposer.dispose();
    }

    start_rendering(): void {
        this.renderer.start_rendering();
    }

    stop_rendering(): void {
        this.renderer.stop_rendering();
    }

    set_size(width: number, height: number): void {
        this.renderer.set_size(width, height);
    }

    private nj_object_or_xvm_changed = (): void => {
        this.renderer.destroy_scene();

        const nj_object = this.store.current_nj_object.val;

        if (nj_object) {
            // Convert textures and geometry.
            const node = ninja_object_to_node(nj_object);
            this.renderer.scene.root_node.add_child(node);

            this.renderer.scene.traverse(node => {
                node.mesh?.upload(this.renderer.gfx);
            }, undefined);
        }

        this.renderer.schedule_render();
    };
}
