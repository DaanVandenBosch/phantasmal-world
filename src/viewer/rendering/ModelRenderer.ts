import {
    AnimationClip,
    AnimationMixer,
    Clock,
    DoubleSide,
    MeshLambertMaterial,
    Object3D,
    PerspectiveCamera,
    SkeletonHelper,
    SkinnedMesh,
    Vector3,
} from "three";
import { Disposable } from "../../core/observable/Disposable";
import { NjMotion } from "../../core/data_formats/parsing/ninja/motion";
import { xvr_texture_to_three_texture } from "../../core/rendering/conversion/ninja_textures";
import { create_mesh } from "../../core/rendering/conversion/create_mesh";
import { ninja_object_to_buffer_geometry } from "../../core/rendering/conversion/ninja_three_geometry";
import {
    create_animation_clip,
    PSO_FRAME_RATE,
} from "../../core/rendering/conversion/ninja_animation";
import { DisposableThreeRenderer, ThreeRenderer } from "../../core/rendering/ThreeRenderer";
import { Disposer } from "../../core/observable/Disposer";
import { ChangeEvent } from "../../core/observable/Observable";
import { LogManager } from "../../core/Logger";
import { ModelStore } from "../stores/ModelStore";
import { CharacterClassModel } from "../model/CharacterClassModel";

const logger = LogManager.get("viewer/rendering/ModelRenderer");

const DEFAULT_MATERIAL = new MeshLambertMaterial({
    color: 0xffffff,
    side: DoubleSide,
});
const DEFAULT_SKINNED_MATERIAL = new MeshLambertMaterial({
    skinning: true,
    color: 0xffffff,
    side: DoubleSide,
});
const CAMERA_POSITION = Object.freeze(new Vector3(0, 10, 20));
const CAMERA_LOOK_AT = Object.freeze(new Vector3(0, 0, 0));

export class ModelRenderer extends ThreeRenderer implements Disposable {
    private readonly disposer = new Disposer();
    private readonly clock = new Clock();
    private character_class_active: boolean;
    private mesh?: Object3D;
    private skeleton_helper?: SkeletonHelper;
    private animation?: {
        mixer: AnimationMixer;
        clip: AnimationClip;
    };
    private update_animation_time = true;

    readonly camera = new PerspectiveCamera(75, 1, 1, 200);

    constructor(private readonly store: ModelStore, three_renderer: DisposableThreeRenderer) {
        super(three_renderer);

        this.character_class_active = store.current_character_class.val != undefined;

        this.disposer.add_all(
            store.current_character_class.observe(this.current_character_class_changed),
            store.current_nj_object.observe(this.nj_object_or_xvm_changed),
            store.current_textures.observe(this.nj_object_or_xvm_changed),
            store.current_nj_motion.observe(this.nj_motion_changed),
            store.show_skeleton.observe(this.show_skeleton_changed),
            store.animation_playing.observe(this.animation_playing_changed),
            store.animation_frame_rate.observe(this.animation_frame_rate_changed),
            store.animation_frame.observe(this.animation_frame_changed),
        );

        this.init_camera_controls();
        this.reset_camera(CAMERA_POSITION, CAMERA_LOOK_AT);
    }

    set_size(width: number, height: number): void {
        this.camera.aspect = width / height;
        this.camera.updateProjectionMatrix();
        super.set_size(width, height);
    }

    dispose(): void {
        super.dispose();
        this.disposer.dispose();
    }

    protected render(): void {
        if (this.animation) {
            this.animation.mixer.update(this.clock.getDelta());
        }

        this.light_holder.quaternion.copy(this.camera.quaternion);
        super.render();

        if (this.animation && !this.animation.mixer.clipAction(this.animation.clip).paused) {
            this.update_animation_frame();
            this.schedule_render();
        }
    }

    private current_character_class_changed = (
        change: ChangeEvent<CharacterClassModel | undefined>,
    ): void => {
        const character_class_active = change.value != undefined;

        if (this.character_class_active !== character_class_active) {
            this.reset_camera(CAMERA_POSITION, CAMERA_LOOK_AT);
        }

        this.character_class_active = character_class_active;
    };

    private nj_object_or_xvm_changed = (): void => {
        if (this.mesh) {
            this.scene.remove(this.mesh);
            this.mesh = undefined;
            this.scene.remove(this.skeleton_helper!);
            this.skeleton_helper = undefined;
        }

        const nj_object = this.store.current_nj_object.val;

        if (nj_object) {
            // Stop animation and store animation time.
            let animation_time: number | undefined;

            if (this.animation) {
                const mixer = this.animation.mixer;
                animation_time = mixer.existingAction(this.animation.clip)?.time;
                mixer.stopAllAction();
                mixer.uncacheAction(this.animation.clip);
                this.animation = undefined;
            }

            // Convert textures and geometry.
            const textures = this.store.current_textures.val.map(tex => {
                if (tex) {
                    try {
                        return xvr_texture_to_three_texture(tex);
                    } catch (e) {
                        logger.error("Couldn't convert XVR texture.", e);
                    }
                }

                return undefined;
            });

            const geometry = ninja_object_to_buffer_geometry(nj_object);
            const has_skeleton = geometry.getAttribute("skinIndex") != undefined;

            this.mesh = create_mesh(
                geometry,
                textures,
                has_skeleton ? DEFAULT_SKINNED_MATERIAL : DEFAULT_MATERIAL,
                has_skeleton,
            );

            // Make sure we rotate around the center of the model instead of its origin.
            const bb = geometry.boundingBox!;
            const height = bb.max.y - bb.min.y;
            this.mesh.translateY(-height / 2 - bb.min.y);

            this.scene.add(this.mesh);

            // Add skeleton.
            this.skeleton_helper = new SkeletonHelper(this.mesh);
            this.skeleton_helper.visible = this.store.show_skeleton.val;
            (this.skeleton_helper.material as any).linewidth = 3;
            this.scene.add(this.skeleton_helper);

            // Create a new animation mixer and clip.
            const nj_motion = this.store.current_nj_motion.val;

            if (nj_motion) {
                const mixer = new AnimationMixer(this.mesh);
                mixer.timeScale = this.store.animation_frame_rate.val / PSO_FRAME_RATE;

                const clip = create_animation_clip(nj_object, nj_motion);

                this.animation = { mixer, clip };

                const action = mixer.clipAction(clip, this.mesh);
                action.time = animation_time ?? 0;
                action.play();
            }
        }

        this.schedule_render();
    };

    private nj_motion_changed = ({ value: nj_motion }: ChangeEvent<NjMotion | undefined>): void => {
        let mixer: AnimationMixer | undefined;

        if (this.animation) {
            this.animation.mixer.stopAllAction();
            this.animation.mixer.uncacheAction(this.animation.clip);
            mixer = this.animation.mixer;
        }

        const nj_object = this.store.current_nj_object.val;

        if (!this.mesh || !(this.mesh instanceof SkinnedMesh) || !nj_motion || !nj_object) return;

        if (!mixer) {
            mixer = new AnimationMixer(this.mesh);
        }

        const clip = create_animation_clip(nj_object, nj_motion);

        this.animation = { mixer, clip };

        this.clock.start();
        mixer.clipAction(clip).play();
        this.schedule_render();
    };

    private show_skeleton_changed = ({ value: show_skeleton }: ChangeEvent<boolean>): void => {
        if (this.skeleton_helper) {
            this.skeleton_helper.visible = show_skeleton;
            this.schedule_render();
        }
    };

    private animation_playing_changed = ({ value: playing }: ChangeEvent<boolean>): void => {
        if (this.animation) {
            this.animation.mixer.clipAction(this.animation.clip).paused = !playing;

            if (playing) {
                this.clock.start();
                this.schedule_render();
            } else {
                this.clock.stop();
            }
        }
    };

    private animation_frame_rate_changed = ({ value: frame_rate }: ChangeEvent<number>): void => {
        if (this.animation) {
            this.animation.mixer.timeScale = frame_rate / PSO_FRAME_RATE;
        }
    };

    private animation_frame_changed = ({ value: frame }: ChangeEvent<number>): void => {
        const nj_motion = this.store.current_nj_motion.val;

        if (this.animation && nj_motion) {
            const frame_count = nj_motion.frame_count;
            if (frame > frame_count) frame = 1;
            if (frame < 1) frame = frame_count;

            if (this.update_animation_time) {
                this.animation.mixer.clipAction(this.animation.clip).time =
                    (frame - 1) / PSO_FRAME_RATE;
            }

            this.schedule_render();
        }
    };

    private update_animation_frame(): void {
        if (this.animation) {
            const action = this.animation.mixer.clipAction(this.animation.clip);

            if (!action.paused) {
                this.update_animation_time = false;
                this.store.set_animation_frame(action.time * PSO_FRAME_RATE + 1);
                this.update_animation_time = true;
            }
        }
    }
}
