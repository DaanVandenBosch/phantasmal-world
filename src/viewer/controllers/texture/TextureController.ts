import { Controller } from "../../../core/controllers/Controller";
import { filename_extension } from "../../../core/util";
import { read_file } from "../../../core/files";
import { is_xvm, parse_xvm, XvrTexture } from "../../../core/data_formats/parsing/ninja/texture";
import { ArrayBufferCursor } from "../../../core/data_formats/block/cursor/ArrayBufferCursor";
import { Endianness } from "../../../core/data_formats/block/Endianness";
import { parse_afs } from "../../../core/data_formats/parsing/afs";
import { LogManager } from "../../../core/Logger";
import { WritableListProperty } from "../../../core/observable/property/list/WritableListProperty";
import { list_property, property } from "../../../core/observable";
import { ListProperty } from "../../../core/observable/property/list/ListProperty";
import { prs_decompress } from "../../../core/data_formats/compression/prs/decompress";
import { failure, Result, result_builder } from "../../../core/Result";
import { Severity } from "../../../core/Severity";
import { Property } from "../../../core/observable/property/Property";
import { WritableProperty } from "../../../core/observable/property/WritableProperty";

const logger = LogManager.get("viewer/controllers/TextureController");

export class TextureController extends Controller {
    private readonly _textures: WritableListProperty<XvrTexture> = list_property();
    private readonly _result_dialog_visible = property(false);
    private readonly _result: WritableProperty<Result<unknown> | undefined> = property(undefined);
    private readonly _result_problems_message = property("");
    private readonly _result_error_message = property("");

    readonly textures: ListProperty<XvrTexture> = this._textures;
    readonly result_dialog_visible: Property<boolean> = this._result_dialog_visible;
    readonly result: Property<Result<unknown> | undefined> = this._result;
    readonly result_problems_message: Property<string> = this._result_problems_message;
    readonly result_error_message: Property<string> = this._result_error_message;

    load_file = async (file: File): Promise<void> => {
        this._result_problems_message.val = `Encountered some problems while opening "${file.name}".`;
        this._result_error_message.val = `Couldn't open "${file.name}".`;

        try {
            const ext = filename_extension(file.name).toLowerCase();
            const buffer = await read_file(file);
            const cursor = new ArrayBufferCursor(buffer, Endianness.Little);

            if (ext === "xvm") {
                const xvm_result = parse_xvm(cursor);
                this.set_result(xvm_result);

                if (xvm_result.success) {
                    this._textures.val = xvm_result.value.textures;
                }
            } else if (ext === "afs") {
                const rb = result_builder(logger);
                const afs_result = parse_afs(cursor);
                rb.add_result(afs_result);

                if (!afs_result.success) {
                    this.set_result(rb.failure());
                } else {
                    const textures: XvrTexture[] = afs_result.value.flatMap(file => {
                        const cursor = new ArrayBufferCursor(file, Endianness.Little);

                        if (is_xvm(cursor)) {
                            const xvm_result = parse_xvm(cursor);
                            rb.add_result(xvm_result);
                            return xvm_result.value?.textures ?? [];
                        } else {
                            const xvm_result = parse_xvm(prs_decompress(cursor.seek_start(0)));
                            rb.add_result(xvm_result);
                            return xvm_result.value?.textures ?? [];
                        }
                    });

                    if (textures.length) {
                        this.set_result(rb.success(textures));
                    } else {
                        this.set_result(rb.failure());
                    }

                    this._textures.val = textures;
                }
            } else {
                logger.debug(`Unsupported file extension in filename "${file.name}".`);
                this.set_result(
                    failure([{ severity: Severity.Error, ui_message: "Unsupported file type." }]),
                );
            }
        } catch (e) {
            logger.error("Couldn't read file.", e);
            this.set_result(failure());
        }
    };

    dismiss_result_dialog = (): void => {
        this._result_dialog_visible.val = false;
    };

    private set_result(result: Result<unknown>): void {
        this._result.val = result;

        if (result.problems.length) {
            this._result_dialog_visible.val = true;
        }
    }
}
