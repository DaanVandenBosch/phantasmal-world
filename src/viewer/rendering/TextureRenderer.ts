import {
    Mesh,
    MeshBasicMaterial,
    OrthographicCamera,
    PlaneGeometry,
    Texture,
    Vector2,
    Vector3,
} from "three";
import { Disposable } from "../../core/observable/Disposable";
import { Renderer } from "../../core/rendering/Renderer";
import { Disposer } from "../../core/observable/Disposer";
import { Xvm } from "../../core/data_formats/parsing/ninja/texture";
import { xvm_texture_to_texture } from "../../core/rendering/conversion/ninja_textures";
import { TextureStore } from "../stores/TextureStore";
import Logger = require("js-logger");

const logger = Logger.get("viewer/rendering/TextureRenderer");

export class TextureRenderer extends Renderer implements Disposable {
    private readonly disposer = new Disposer();
    private readonly quad_meshes: Mesh[] = [];

    readonly camera = new OrthographicCamera(-400, 400, 300, -300, 1, 10);

    constructor(texture_store: TextureStore) {
        super();

        this.disposer.add_all(
            texture_store.current_xvm.observe(({ value: xvm }) => {
                this.scene.remove(...this.quad_meshes);

                if (xvm) {
                    this.render_textures(xvm);
                }

                this.reset_camera(new Vector3(0, 0, 5), new Vector3());
                this.schedule_render();
            }),
        );

        this.init_camera_controls();
        this.controls.dollySpeed = -1;
        this.controls.azimuthRotateSpeed = 0;
        this.controls.polarRotateSpeed = 0;
    }

    set_size(width: number, height: number): void {
        this.camera.left = -Math.floor(width / 2);
        this.camera.right = Math.ceil(width / 2);
        this.camera.top = Math.floor(height / 2);
        this.camera.bottom = -Math.ceil(height / 2);
        this.camera.updateProjectionMatrix();
        super.set_size(width, height);
    }

    dispose(): void {
        super.dispose();
        this.disposer.dispose();
    }

    private render_textures(xvm: Xvm): void {
        let total_width = 10 * (xvm.textures.length - 1); // 10px spacing between textures.
        let total_height = 0;

        for (const tex of xvm.textures) {
            total_width += tex.width;
            total_height = Math.max(total_height, tex.height);
        }

        let x = -Math.floor(total_width / 2);
        const y = -Math.floor(total_height / 2);

        for (const tex of xvm.textures) {
            let tex_3js: Texture | undefined;

            try {
                tex_3js = xvm_texture_to_texture(tex);
            } catch (e) {
                logger.warn("Couldn't convert XVM texture.", e);
            }

            const quad_mesh = new Mesh(
                this.create_quad(
                    x,
                    y + Math.floor((total_height - tex.height) / 2),
                    tex.width,
                    tex.height,
                ),
                tex_3js
                    ? new MeshBasicMaterial({
                          map: tex_3js,
                          transparent: true,
                      })
                    : new MeshBasicMaterial({
                          color: 0xff00ff,
                      }),
            );

            this.quad_meshes.push(quad_mesh);
            this.scene.add(quad_mesh);

            x += 10 + tex.width;
        }
    }

    private create_quad(x: number, y: number, width: number, height: number): PlaneGeometry {
        const quad = new PlaneGeometry(width, height, 1, 1);
        quad.faceVertexUvs = [
            [
                [new Vector2(0, 0), new Vector2(0, 1), new Vector2(1, 0)],
                [new Vector2(0, 1), new Vector2(1, 1), new Vector2(1, 0)],
            ],
        ];
        quad.translate(x + width / 2, y + height / 2, -5);
        return quad;
    }
}
