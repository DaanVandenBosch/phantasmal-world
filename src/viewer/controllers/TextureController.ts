import { Controller } from "../../core/controllers/Controller";
import { filename_extension } from "../../core/util";
import { read_file } from "../../core/files";
import { is_xvm, parse_xvm, XvrTexture } from "../../core/data_formats/parsing/ninja/texture";
import { ArrayBufferCursor } from "../../core/data_formats/cursor/ArrayBufferCursor";
import { Endianness } from "../../core/data_formats/Endianness";
import { parse_afs } from "../../core/data_formats/parsing/afs";
import { LogManager } from "../../core/Logger";
import { WritableListProperty } from "../../core/observable/property/list/WritableListProperty";
import { list_property } from "../../core/observable";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { prs_decompress } from "../../core/data_formats/compression/prs/decompress";
import { failure, Result, result_builder } from "../../core/Result";
import { show_result_popup } from "../../core/gui/ResultPopup";
import { Severity } from "../../core/Severity";

const logger = LogManager.get("viewer/controllers/TextureController");

export class TextureController extends Controller {
    private readonly _textures: WritableListProperty<XvrTexture> = list_property();
    readonly textures: ListProperty<XvrTexture> = this._textures;

    load_file = async (file: File): Promise<void> => {
        let result: Result<unknown>;

        try {
            const ext = filename_extension(file.name).toLowerCase();
            const buffer = await read_file(file);
            const cursor = new ArrayBufferCursor(buffer, Endianness.Little);

            if (ext === "xvm") {
                const xvm_result = (result = parse_xvm(cursor));

                if (xvm_result.success) {
                    this._textures.val = xvm_result.value.textures;
                }
            } else if (ext === "afs") {
                const rb = result_builder(logger);
                const afs_result = parse_afs(cursor);
                rb.add_result(afs_result);

                if (!afs_result.success) {
                    result = rb.failure();
                } else {
                    const textures: XvrTexture[] = afs_result.value.flatMap(file => {
                        const cursor = new ArrayBufferCursor(file, Endianness.Little);

                        if (is_xvm(cursor)) {
                            const xvm_result = parse_xvm(cursor);
                            rb.add_result(xvm_result);
                            return xvm_result.value?.textures ?? [];
                        } else {
                            const xvm_result = parse_xvm(prs_decompress(cursor.seek_start(0)));
                            rb.add_result(xvm_result);
                            return xvm_result.value?.textures ?? [];
                        }
                    });

                    if (textures.length) {
                        result = rb.success(textures);
                    } else {
                        result = rb.failure();
                    }

                    this._textures.val = textures;
                }
            } else {
                logger.debug(`Unsupported file extension in filename "${file.name}".`);
                result = failure([
                    { severity: Severity.Error, ui_message: "Unsupported file type." },
                ]);
            }
        } catch (e) {
            logger.error("Couldn't read file.", e);
            result = failure();
        }

        show_result_popup(
            result,
            `Encountered some problems while opening "${file.name}".`,
            `Couldn't open "${file.name}".`,
        );
    };
}
