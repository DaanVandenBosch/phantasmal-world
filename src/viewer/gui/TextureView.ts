import { div, Icon } from "../../core/gui/dom";
import { FileButton } from "../../core/gui/FileButton";
import { ToolBar } from "../../core/gui/ToolBar";
import { RendererWidget } from "../../core/gui/RendererWidget";
import { TextureRenderer } from "../rendering/TextureRenderer";
import { ResizableView } from "../../core/gui/ResizableView";
import { TextureController } from "../controllers/TextureController";

export class TextureView extends ResizableView {
    readonly element = div({ className: "viewer_TextureView" });

    private readonly open_file_button = new FileButton("Open file...", {
        icon_left: Icon.File,
        accept: ".afs, .xvm",
    });

    private readonly tool_bar = this.add(new ToolBar(this.open_file_button));

    private readonly renderer_view: RendererWidget;

    constructor(ctrl: TextureController, renderer: TextureRenderer) {
        super();

        this.renderer_view = this.add(new RendererWidget(renderer));

        this.element.append(this.tool_bar.element, this.renderer_view.element);

        this.disposables(
            this.open_file_button.files.observe(({ value: files }) => {
                if (files.length) ctrl.load_file(files[0]);
            }),
        );

        this.finalize_construction();
    }

    activate(): void {
        this.renderer_view.start_rendering();
        super.activate();
    }

    deactivate(): void {
        super.deactivate();
        this.renderer_view.stop_rendering();
    }

    resize(width: number, height: number): this {
        super.resize(width, height);

        this.renderer_view.resize(width, Math.max(0, height - this.tool_bar.height));

        return this;
    }
}
