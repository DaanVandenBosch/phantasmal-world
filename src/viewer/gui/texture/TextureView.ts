import { div, Icon } from "../../../core/gui/dom";
import { FileButton } from "../../../core/gui/FileButton";
import { ToolBar } from "../../../core/gui/ToolBar";
import { RendererWidget } from "../../../core/gui/RendererWidget";
import { ResizableView } from "../../../core/gui/ResizableView";
import { TextureController } from "../../controllers/texture/TextureController";
import { ResultDialog } from "../../../core/gui/ResultDialog";
import { Renderer } from "../../../core/rendering/Renderer";

export class TextureView extends ResizableView {
    readonly element = div({ className: "viewer_TextureView" });

    private readonly open_file_button = new FileButton({
        icon_left: Icon.File,
        text: "Open file...",
        accept: ".afs, .xvm",
    });

    private readonly tool_bar = this.add(new ToolBar(this.open_file_button));

    private readonly renderer_widget: RendererWidget;

    constructor(ctrl: TextureController, renderer: Renderer) {
        super();

        this.renderer_widget = this.add(new RendererWidget(renderer));

        this.element.append(this.tool_bar.element, this.renderer_widget.element);

        const dialog = this.disposable(
            new ResultDialog({
                visible: ctrl.result_dialog_visible,
                result: ctrl.result,
                problems_message: ctrl.result_problems_message,
                error_message: ctrl.result_error_message,
            }),
        );

        this.disposables(
            this.open_file_button.files.observe(({ value: files }) => {
                if (files.length) {
                    ctrl.load_file(files[0]);
                }
            }),

            dialog.ondismiss.observe(ctrl.dismiss_result_dialog),
        );

        this.finalize_construction(TextureView);
    }

    resize(width: number, height: number): this {
        super.resize(width, height);

        this.renderer_widget.resize(width, Math.max(0, height - this.tool_bar.height));

        return this;
    }
}
