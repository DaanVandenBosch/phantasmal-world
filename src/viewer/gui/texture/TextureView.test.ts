import { TextureView } from "./TextureView";
import { with_disposer } from "../../../../test/src/core/observables/disposable_helpers";
import { TextureController } from "../../controllers/texture/TextureController";
import { TextureRenderer } from "../../rendering/TextureRenderer";
import { STUB_RENDERER } from "../../../../test/src/core/rendering/StubRenderer";

test("Renders correctly without textures.", () =>
    with_disposer(disposer => {
        const ctrl = disposer.add(new TextureController());
        const view = disposer.add(new TextureView(ctrl, new TextureRenderer(ctrl, STUB_RENDERER)));

        expect(view.element).toMatchSnapshot("Should render a toolbar and a renderer widget.");
    }));
