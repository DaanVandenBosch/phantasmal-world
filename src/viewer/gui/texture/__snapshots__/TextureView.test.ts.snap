// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly without textures.: Should render a toolbar and a renderer widget. 1`] = `
<div
  class="viewer_TextureView"
>
  <div
    class="core_ToolBar"
    style="height: 33px;"
  >
    <button
      class="core_Button core_FileButton"
    >
      <span
        class="core_Button_inner"
      >
        <span
          class="core_Button_left"
        >
          <span>
            <span
              class="fas fa-file"
            />
          </span>
        </span>
        <span
          class="core_Button_center"
        >
          Open file...
        </span>
      </span>
    </button>
  </div>
  <div
    class="core_RendererWidget"
  >
    <canvas
      height="600"
      width="800"
    />
  </div>
</div>
`;
